node/network/approval-distribution/src/lib.rs

```diff
@@ -0,0 +1,826 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! [`ApprovalDistributionSubsystem`] implementation.
+//!
+//! https://w3f.github.io/parachain-implementers-guide/node/approval/approval-distribution.html
+
+#![warn(missing_docs)]
+
+#[cfg(test)]
+mod tests;
+
+
+use std::collections::{BTreeMap, HashMap, HashSet, hash_map};
+use futures::{channel::oneshot, FutureExt as _};
+use polkadot_primitives::v1::{
+	Hash, BlockNumber, ValidatorIndex, ValidatorSignature, CandidateIndex,
+};
+use polkadot_node_primitives::{
+	approval::{AssignmentCert, BlockApprovalMeta, IndirectSignedApprovalVote, IndirectAssignmentCert},
+};
+use polkadot_node_subsystem::{
+	messages::{
+		AllMessages, ApprovalDistributionMessage, ApprovalVotingMessage, NetworkBridgeMessage,
+		AssignmentCheckResult, ApprovalCheckResult,
+	},
+	ActiveLeavesUpdate, FromOverseer, OverseerSignal, SpawnedSubsystem, Subsystem, SubsystemContext,
+};
+use polkadot_node_subsystem_util::metrics::{self, prometheus};
+use polkadot_node_network_protocol::{
+	PeerId, View, NetworkBridgeEvent, v1 as protocol_v1, ReputationChange as Rep,
+};
+
+const LOG_TARGET: &str = "approval_distribution";
+
+// TODO: justify the numbers:
```

Comment:
The problem with this, is that the numbers must be consistent not only within a module, but across all subsystems in polkadot and substrate. Otherwise it might be possible to be honest to one subsystem, gaining reputation disproportionate to the loss of an attack on another subsystem.

To mitigate this, would be good to have a section in the guide with the recommendation of the rewards having the same order of magnitude across all subsystems, which is one or two less of the punishment points.
WDYT?

Reply:
Sure, clarifying that in the guide is a good idea.

Reply:
cc @eskimor - this is one of the nuisances to watch out for with our network stack

Reply:
Hmm, shouldn't we then define those rewards/costs in a single place - for all subsystems? Or, alternatively have constants defined in one place like 'COST_MINOR_ANNOY', 'COST_ANNOY', 'COST_MALICIOUS' ... which should be used by subsystems to define their own costs, instead of raw numbers? This way it would be harder to break the consistency by accident.

I am always a fan of making intentions explicit in the code. Ideally even add a newtype wrapper, so subsystems must use those predefined values instead of making numbers up on their own.